### PR TITLE
refactor: unify instruction parsing across tx page and inspector

### PR DESCRIPTION
**Pull request by askov**
## Description

Unifies the two instruction-parsing pipelines (tx page + inspector) behind a single shared `instruction-parser` entity. Each program now has one parser slice that owns its canonical `SliceParsed` shape; both surfaces feed through one dispatcher and the tx page's `InstructionCard` render pipeline.

- **`app/entities/instruction-parser/`** — shared dispatcher, provider, compat wrap (`toParsedInstruction`), and a contract test.
- **Per-program feature slices** — `instruction-system`, `instruction-token`, `instruction-token-2022`, `instruction-associated-token`, and `mpl-token-metadata`, each exposing `fromParsed` (RPC) and `fromTransaction` (raw bytes).
- **Inspector** becomes a thin wrapper that decodes raw `TransactionInstruction`s through the same slices the tx page already trusts via RPC pre-parsing — instead of the old `into-parsed-data.ts` adapter (which covered only five programs, mutated its input, shadowed `PROGRAM_INFO_BY_ID`, and returned `any`).
- **Contract test** keeps the byte-parse and RPC-parse pipelines from silently drifting.

This is Phase 1 of the unification (see `docs/instruction-parser-unification.md`); the `toParsedInstruction` compat shim is transitional and removed in a later phase once cards consume `SliceParsed` directly.

## Type of change

-   [x] Other (please describe): Refactoring — no user-facing behaviour change; unifies two parsing pipelines

## Screenshots

No visual change — this is a pure refactor. Tx-page and inspector parity verified manually (see Testing below).

## Testing

Unit + contract tests pass locally (`pnpm test:ci`). Manual verification steps and full link list in `docs/manual-verification.md`. Key parity checks (open tx-page and inspector side-by-side — fields MUST match):

- System Transfer — [tx](http://localhost:3000/tx/E7zy6gPCa3XRy5ykdb2Pt7nb2RHKPWhtnQ5jWgJat4M25jGy3PFAcowtBiby4YyvFRgipu1Js6iX9NpU1WNbFNJ) · [inspector](http://localhost:3000/tx/E7zy6gPCa3XRy5ykdb2Pt7nb2RHKPWhtnQ5jWgJat4M25jGy3PFAcowtBiby4YyvFRgipu1Js6iX9NpU1WNbFNJ/inspect)
- Associated Token Create (idempotent) — [tx](http://localhost:3000/tx/Nwa7VZEfhbPV2MmW4dQNebj6Pw1TrJH9j2G3pL7Vt4Hr2rTzyEPdh1LBd4SKtzEBhC1MKMDEeaQuz96FZzQSyYk) · [inspector](http://localhost:3000/tx/Nwa7VZEfhbPV2MmW4dQNebj6Pw1TrJH9j2G3pL7Vt4Hr2rTzyEPdh1LBd4SKtzEBhC1MKMDEeaQuz96FZzQSyYk/inspect)
- Associated Token Create (with inner cards) — [tx](http://localhost:3000/tx/4F3kU8ssrWwn7t4YCyVNNbxbP2D3YgHmPL6efGycLoxM3aRFo2Fi1D6UgzxcK8ZMTmgmBqavUtchA7S4dewSbnfs) · [inspector](http://localhost:3000/tx/4F3kU8ssrWwn7t4YCyVNNbxbP2D3YgHmPL6efGycLoxM3aRFo2Fi1D6UgzxcK8ZMTmgmBqavUtchA7S4dewSbnfs/inspect)
- Token TransferChecked — [tx](http://localhost:3000/tx/X8DpDCrFN8Ljsh4wcCKLvrsBXAeuxCt8xj4re6dZkESYT95QiVraC9Q7ra6afekg9HvQt4B78K8VNgcQrzZqBb5) · [inspector](http://localhost:3000/tx/X8DpDCrFN8Ljsh4wcCKLvrsBXAeuxCt8xj4re6dZkESYT95QiVraC9Q7ra6afekg9HvQt4B78K8VNgcQrzZqBb5/inspect)

## Related Issues
TBD

## Checklist

-   [x] My code follows the project's style guidelines
-   [x] I have added tests that prove my fix/feature works
-   [x] All tests pass locally and in CI
-   [x] I have updated documentation as needed
-   [ ] I have run `build:info` script to update build information
-   [x] CI/CD checks pass

## Additional Notes

The `instruction-parser` spec lives at `openspec/specs/instruction-parser/spec.md`. Phase 1 only — later phases migrate cards off the compat shim.

### CODE REVIEW

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
+# instruction-parser
+
+## Purpose
+
+One normalisation layer for Solana instruction data, shared by `/tx/[signature]` and `/tx/(inspector)/inspector`. Both surfaces fold their input through the same per-program slice and render the same `*DetailsCard` components. Adding a program is one slice plus one line in the shared dispatcher; a contract test pins the two paths to the same canonical shape so they cannot silently disagree.
+
+## Architecture
+
+```
+  Inspector                              TX page
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: The inspector could support both:
- loading transaction by signature
- using a versioned message in wire format

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                    <DetailsCard ix={…} />
+```
+
+`SliceParsed` is the slice-owned canonical shape (typically a discriminated union). The compat layer at `app/entities/instruction-parser/model/compat.ts` holds every shim that lets inspector input flow through tx-page-designed cards: `toParsedInstruction` wraps `SliceParsed` back to `ParsedInstruction`, and `toParsedTransaction` builds a synthetic `ParsedTransaction` around a single instruction for cards that take a `tx` prop. Both disappear in one deletion when cards consume `SliceParsed` directly.
+
+## ADDED Requirements
+
+### Requirement: Per-program parser slices
+
+Each supported Solana program SHALL be implemented as a feature slice at `app/features/instruction-<program>/` exposing an `InstructionParser<P>` value, where `P extends ParsedInstructionInfo` is the slice's canonical shape.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: 👍

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+The parser SHALL set `programId` (base58) and `programLabel` matching the RPC `program` field. It SHALL implement `fromTransaction(ix: KitInstruction): P | undefined` and MAY implement `fromParsed(ix: ParsedInstruction): P | undefined`. Slice internals MUST NOT import `@solana/web3.js` types — the bridge is `KitInstruction` from `@/app/shared/lib/web3js-compat`.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:57]: I think that we should also mention the LegacyTransaction variant. Not sure we could easily migrate to the kit at once.

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: RPC-parsed program publishes both paths
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:61]: Do you think it is worth adding a method, like `canHandle*` to determine whether we can use one variant or another?

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+### Requirement: Dispatcher contract
+
+The entity at `app/entities/instruction-parser/` SHALL expose a factory `createInstructionParserDispatcher(parsers)` that throws on duplicate `programId`. The returned dispatcher SHALL expose three methods:
+
+- `fromTransactionInstruction(ix): DispatchResult | undefined`
+- `fromParsedInstruction(ix): ParsedInstruction`
+- `getInstructionParser(programId): InstructionParser | undefined`
+
+`DispatchResult` is `ParsedInstruction | DispatchUnknown` where `DispatchUnknown = { unknown: true; programLabel: string; programId: PublicKey }`. The dispatcher MUST be pure and MUST NOT hold module-level state. It is delivered to consumers via `InstructionParserProvider` / `useInstructionParser`; the hook MUST throw when used outside the provider.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: I'd expect `ParsedInvalid`, `NotParsable` or similar in conjunction with `ParsedInstruction`.
`ParsedInstruction | DispatchUnknown` look a bit strange together, unrelated even

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+## Why
+
+Every new program or `*DetailsCard` improvement had to land on **both** surfaces or the inspector silently fell behind — and it already had. With no shared contract test, the two pipelines drift apart and the same instruction renders differently depending on which page the user opened.
+
+The goal: **one parser per program, one canonical shape, one place to add a program.** The inspector becomes a thin wrapper that feeds raw `TransactionInstruction`s through the same parsers the tx page already trusts via RPC pre-parsing. A contract test pins both paths to the same shape so they cannot disagree.
+
+### Alternatives considered
+
+- **Inspector mimics RPC's `ParsedInstruction` shape (original draft).** Rejected as asymmetric — it privileges the RPC shape and bakes web3.js's `parsed.info` structure into every slice. Chosen instead: each slice owns a canonical, typed `SliceParsed` (usually a discriminated union); **both** surfaces converge on it. RPC is not "the default"; the slice is.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/proposal.md:17]: 👍

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+- **Per-route dispatchers wired into each page.** Rejected in favor of a single shared dispatcher (`app/tx/instruction-parser-dispatcher.ts`) delivered once via `InstructionParserProvider` in `app/tx/layout.tsx`, the common ancestor of all `/tx` routes. Adding a program touches one list; every route inherits it.
+- **Module-level global registry / a top-level `instruction-parsers/` barrel.** Rejected — the barrel violated FSD slice isolation and a global registry holds hidden state. Chosen instead: a pure `createInstructionParserDispatcher(parsers)` factory delivered through React Context.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/proposal.md:19]: 👍

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+- **Migrate all `*DetailsCard` components to consume `SliceParsed` now.** Deferred to a future change. A transitional compat wrap (`toParsedInstruction`) keeps the ~15-20 cards and their superstruct validators unchanged so this change stays additive; collapsing the render pipeline and deleting the wrap is out of scope here.
+- **One Token slice covering Token + Token-2022.** Rejected — distinct `programId`s, distinct `@solana-program/*` packages, different instruction supersets. One slice per `programId`.
+
+## What Changes
+
+This change builds the shared foundation and migrates the five programs the inspector already parsed. It is intentionally additive — no card or validator changes, no behaviour change.
+
+- Add the **`instruction-parser` entity** (`app/entities/instruction-parser/`): the `InstructionParser<P>` / `ParsedInstructionInfo` / `InstructionParserDispatcher` contract, the `createInstructionParserDispatcher` factory (throws on duplicate `programId`, holds no module state), the `InstructionParserProvider` / `useInstructionParser` Context delivery (hook throws outside a provider), and the transitional compat layer (`toParsedInstruction`, `toParsedTransaction`).
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/proposal.md:27]: I'd suggest another name. We may have a lot of different functionality with `instruction-something` on the entity and feature level.
As we gonna utilize buffer data, maybe `decode` is a better term, even if the solana-program clients have methods to "parse" instructions.

Comment (rogaldh) [reply openspec/changes/unify-instruction-parsing/proposal.md:27]: I mostly worried about the features. Pretty sure there will be a lot of those. `decode-instruction-token-2022` seems to me like a better variant

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+- **Wire both surfaces through one shared dispatcher** at `app/tx/instruction-parser-dispatcher.ts`, provided once in `app/tx/layout.tsx`. The tx page normalises RPC input via `dispatcher.fromParsedInstruction(ix)` (pass-through for unsliced programs); the inspector decodes via `dispatcher.fromTransactionInstruction(ix)`.
+- **Delete `app/components/inspector/into-parsed-data.ts`** and its ad-hoc parsers/registry.
+- Add a **cross-pipeline contract test** asserting the byte-parsed and RPC-parsed paths produce equivalent, validator-satisfying `parsed.info` for the same logical instruction; it must extend by one assertion per new slice.
+- The capability spec ships as a delta at `specs/instruction-parser/spec.md` within this change (optional artifact, included because the contract is non-trivial).
+
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/proposal.md:34]: Let's also keep a note how we gonna treat instructions with inner_instructions. Explorer has lack of support these on the inspector page.

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+**Out of scope** (deliberately not in this change, tracked for follow-ups): inspector coverage for programs the old adapter never handled (Stake, Vote, ALT, BPF, Memo, Compute Budget, Lighthouse, SAS, Anchor/Program Metadata IDL, …); collapsing the two per-surface card switches into one render pipeline; deleting the compat wrap by migrating cards to consume `SliceParsed` directly.
+
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/proposal.md:36]: Explorer also displays instructions for Anchor and Squads and we should have a plan, how to handle these

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: Inspector synthetic tx comes from the compat layer
+
+- **WHEN** the inspector renders a card that takes a `tx: ParsedTransaction` prop
+- **THEN** the tx MUST be built via `toParsedTransaction` from the compat layer
+- **AND** no other module in the codebase MUST fabricate a synthetic `ParsedTransaction`
+
+### Requirement: No input mutation
+
+The dispatcher and every slice parser MUST NOT mutate the input `TransactionInstruction` or `ParsedInstruction`. `toKitInstruction` MUST return a fresh object. The Associated Token discriminator workaround MUST construct a local `Uint8Array` via `{ ...ix, data: effective }` rather than writing to `ix.data`.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:143]: Could it be achieved via Readonly or in another way? I think it is worth mentioning.

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: Dispatch leaves input untouched
+
+- **WHEN** a slice parser or the dispatcher processes an instruction
+- **THEN** the input object graph MUST be unchanged after the call
+- **AND** snapshot comparisons before and after MUST find no difference
+
+### Requirement: Cross-pipeline equivalence
+
+A contract test SHALL assert, for every program with both `fromTransaction` and `fromParsed`, that the two paths produce `parsed.info` satisfying the same superstruct validator AND yielding equivalent field values for the same logical instruction. The test lives at `app/tx/__tests__/instruction-parser-contract.spec.ts` (the composition layer, so it may import feature slices without crossing the FSD entity→feature boundary) and MUST extend for every new slice.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: I may be wrong here, but I do remember cases when parsed.info from ParsedInstruciton was different to what decoding from the bytes provided. Some weird cases with multisigs.
I think that `satisfying the same superstruct` might not be feasible in some cases. Maybe to leave a notice here would be good

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: Slice label matches the registry
+
+- **WHEN** a new slice declares `programLabel: '<label>'`
+- **THEN** `PROGRAM_INFO_BY_ID` MUST already publish that label for the same `programId`, or be updated in the same change
+- **AND** `getProgramName()` MUST return that label for the same address
+
+### Requirement: Memoised per-instruction dispatch
+
+The inspector and tx-page `InstructionsSection.tsx` consumers SHALL wrap their dispatcher calls in `useMemo` keyed on the instruction reference, so superstruct validation and `PublicKey` allocations do not re-run on every parent re-render. `useMemo` calls MUST precede any conditional return to respect React's rules-of-hooks.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Looks to me, `useMemo` is too much in-depth piece and out of the scope for that proposal

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: Hover does not re-dispatch
+
+- **WHEN** a parent re-renders for a reason orthogonal to the instruction set (hover, expand, scroll anchor)
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: This one also looks like in-depth and not related to the parsing/decoding story

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+- **THEN** the dispatcher MUST NOT re-execute for the same `ix` reference
+- **AND** no new `PublicKey` instances MUST be allocated for unchanged inputs
+
+### Requirement: Cards consume pre-parsed slice output
+
+A `*DetailsCard` whose program has a registered slice parser MUST consume the dispatcher's already-parsed output rather than re-parsing the raw instruction. `MetaplexTokenMetadataDetailsCard` MUST accept an optional `parsedIx` prop; when provided with a non-empty `parsed.type`, the card MUST reuse it and skip its internal parse.
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:198]: I think, as we have a lot of cards now that expect parsed data. Maybe it is worth mentioning within this spec a way how we gonna support existing cards.
The requirement states that Cards MUST consume, but nothing is said about the cards which "do not have a registered slice".

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: Inspector renders MPL without double-parse
+
+- **WHEN** the inspector routes an MPL instruction to the card with `parsedIx` set
+- **THEN** the card MUST reuse `parsedIx.parsed.{type, info}`
+- **AND** the card MUST NOT call `parseMetaplexTokenMetadataInstruction` again in the same render
+
+#### Scenario: Tx-page predicate path self-parses
+
+- **WHEN** the tx-page predicate branch routes a `PartiallyDecodedInstruction` to the MPL card without `parsedIx`
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:216]: It looks to me that PartiallyDecodedInstruction should be mentioned in the diagram at the top. Explorer really has unions with them here and there and we should clearly state how the new parsing works with them

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 17]
```diff
@@ -0,0 +1,164 @@
+# instruction-parser
+
+## Purpose
+
+One normalisation layer for Solana instruction data, shared by `/tx/[signature]` and `/tx/(inspector)/inspector`. Both surfaces fold their input through the same per-program slice and render the same `*DetailsCard` components. Adding a program is one slice plus one line in the shared dispatcher; a contract test pins the two paths to the same canonical shape so they cannot silently disagree.
+
+## Architecture
+
+```
```
Comment (rogaldh) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: I suggest using mermaidjs to describe the flowcharts. It si widely supported and would be nicely displayed on github.

Sample
```mermaid
flowchart TD
      A[Inspector<br/>TransactionInstruction] --> B[dispatcher.fromTransactionInstruction]
      C[TX page<br/>ParsedInstruction] --> D[dispatcher.fromParsedInstruction]
      B --> E[slice.fromTransaction]
      D --> F[slice.fromParsed]
      E --> G([SliceParsed])
      F --> G
      G --> H[toParsedInstruction<br/>transitional wrap]
      H --> I[ParsedInstruction]
      I --> J[&lt;DetailsCard ix=… /&gt;]
```

Comment (rogaldh) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: OR

```mermaid
 flowchart TD
      A[Inspector<br/>TransactionInstruction] --> B[dispatcher.fromTransactionInstruction]
      C[TX page<br/>ParsedInstruction] --> D[dispatcher.fromParsedInstruction]
      B --> E[slice.fromTransaction]
      D --> F[slice.fromParsed]
      E --> G([SliceParsed])
      F --> G
      G --> H[[toParsedInstruction<br/>transitional wrap]]
      H --> I[ParsedInstruction]
      I --> J[&lt;DetailsCard ix=… /&gt;]
```

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 11 above)
Comment (mikhd) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:143]: May be eslint `"no-param-reassign": ["error", { "props": true }]`?

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+
+#### Scenario: System Transfer parity
+
+- **WHEN** the same `SystemProgram.transfer(...)` is dispatched via `fromTransactionInstruction(rawIx)` and `fromParsedInstruction(rpcIx)`
+- **THEN** both `parsed.info` payloads MUST satisfy `TransferInfo`
+- **AND** the validated `source`, `destination`, and `lamports` MUST be equal across the two paths
+
+### Requirement: Program labels single-source
+
+Each slice's `programLabel` MUST match the entry in `app/utils/programs.ts` (`PROGRAM_INFO_BY_ID`) and the value returned by `getProgramName()` in `app/utils/tx.ts`. Drift between slice labels and the registry would silently route to the wrong card.
```
Comment (mikhd) [inline openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: This statement could be misleading.
It says that label must match `PROGRAM_INFO_BY_ID / getProgramName()`, but, for example, for System program and Token:
- The `PROGRAM_INFO_BY_ID` and `PROGRAM_NAMES` are `SYSTEM = System Program` and `TOKEN = 'Token Program'`.
- The `InstructionParser` `programLabels` are `system` and `spl-token`.

So the `programLabel` from Instruction holds the RPC program field that is used as a guard, for instance, in `parseSystemRpcInstruction` (`if (ix.program !== 'system') return undefined`).

Comment (mikhd) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Also, I think if we'll need some matching between names, it would be awesome if we could use `PROGRAM_NAMES` enum type instead of string to ensure compatibility.

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+- That adapter also mutated its input (`instruction.data`), shadowed `PROGRAM_INFO_BY_ID` with its own 4-entry registry, returned `any`, and routed `CreateAccountWithSeed` through a separate asymmetric path. Even when it produced a `ParsedInstruction`-shaped object, the inspector still rendered through its own card switch rather than the tx page's `InstructionCard`.
+
+## Why
+
+Every new program or `*DetailsCard` improvement had to land on **both** surfaces or the inspector silently fell behind — and it already had. With no shared contract test, the two pipelines drift apart and the same instruction renders differently depending on which page the user opened.
+
+The goal: **one parser per program, one canonical shape, one place to add a program.** The inspector becomes a thin wrapper that feeds raw `TransactionInstruction`s through the same parsers the tx page already trusts via RPC pre-parsing. A contract test pins both paths to the same shape so they cannot disagree.
```
Comment (mikhd) [inline openspec/changes/unify-instruction-parsing/proposal.md:13]: One dispatcher to rule them all! 😄 
Based on Lord of the Rings novel:

```
Three Parsers for the System-kings under the kit,
Seven for the Token-lords in their halls of mints,
Nine for Token-2022 doomed to grow,
One for the Dispatcher on its layout throne
In the Land of /tx where the instructions lie.

One dispatcher to rule them all, one dispatcher to find them,
One dispatcher to bring them all, and to SliceParsed bind them,
In the Land of /tx where the cards no longer drift.
```

File: openspec/changes/unify-instruction-parsing/proposal.md [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+## What Changes
+
+This change builds the shared foundation and migrates the five programs the inspector already parsed. It is intentionally additive — no card or validator changes, no behaviour change.
+
+- Add the **`instruction-parser` entity** (`app/entities/instruction-parser/`): the `InstructionParser<P>` / `ParsedInstructionInfo` / `InstructionParserDispatcher` contract, the `createInstructionParserDispatcher` factory (throws on duplicate `programId`, holds no module state), the `InstructionParserProvider` / `useInstructionParser` Context delivery (hook throws outside a provider), and the transitional compat layer (`toParsedInstruction`, `toParsedTransaction`).
+- Add **per-program feature slices** under `app/features/instruction-*/`: `instruction-system`, `instruction-token`, `instruction-token-2022`, `instruction-associated-token`, and `mpl-token-metadata`. Each publishes `fromTransaction` (raw kit bytes → `SliceParsed`) and, when the RPC pre-parses that program, `fromParsed`. Slice internals use `@solana/kit` and the `KitInstruction` bridge — no `@solana/web3.js` types.
```
Comment (mikhd) [inline openspec/changes/unify-instruction-parsing/proposal.md]: **questions:**
1) Might these `instruction-<program_name>` flood the feature folder eventually? We'll have ~15+, but I guess its fine for FSD?
2) `instruction-<program_name>` name suggests that the scope is the program's instructions.
To understand correctly intention: are we planning to put other instruction functionalities possible in future to the same folder?

May be narrow down naming into something like `ix-decoder-<program_name>` or as @rogaldh suggested `decode-instruction-<program_name>`?

File: app/features/decode-instruction-associated-token/lib/associated-token-parser.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+            default:
+                return undefined;
+        }
+    } catch {
+        return undefined;
+    }
+}
+
+/**
+ * Some clients send the AT Create instruction with empty data instead of the
```
Comment (mikhd) [inline app/features/decode-instruction-associated-token/lib/associated-token-parser.ts:79]: May be worth adding a test for the case with empty data.

```
  test('should parse empty-data instruction as create', () => {
      // Empty data is how some clients send the AT Create instruction.
      // The parser should treat it as Create, not as Unknown.
      const ix = makeAtInstructionWithData(new Uint8Array(0));
      const result = parseAssociatedTokenInstruction(toKitInstruction(ix));
      expect(result?.type).toBe('create');
  });
```

File: app/tx/__tests__/instruction-parser-contract.spec.ts [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+import { Keypair, type ParsedInstruction, SystemProgram, TransactionInstruction } from '@solana/web3.js';
+import { create } from 'superstruct';
+import { describe, expect, test } from 'vitest';
+
+/**
+ * Contract test: both pipelines (inspector via `fromTransactionInstruction`
+ * and tx-page via `fromParsedInstruction`) must produce `parsed.info` payloads
+ * that satisfy the same superstruct validator AND yield equivalent values.
+ * System Transfer is the smallest probe — extend this suite with one
+ * assertion per migrated program as new slices land.
```
Comment (mikhd) [inline app/tx/__tests__/instruction-parser-contract.spec.ts]: **question**: Just to understand the scope of current PR - are we planning to add same parity tests for the rest already added parsers later?

- SPL Token.
- Token-2022.
- Associated Token.
- Mpl token metadata.

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 13 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Agreed - dropped

File: openspec/changes/unify-instruction-parsing/proposal.md (hunk 20 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/proposal.md]: Renamed the four slices to decode-instruction-{system,token,token-2022,associated-token}

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 5 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Renamed `DispatchUnknown → UnparsedInstruction`, so `DispatchResult = ParsedInstruction | UnparsedInstruction` now reads as two parallel outcomes of one decode attempt. I avoided `ParsedInvalid/NotParsable` deliberately — this case isn't "invalid," it's "program recognized, discriminator not decoded" (distinct from undefined = no parser at all), and I didn't want the name to imply the bytes were malformed.

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 18 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: You're right, rewrote it as "Program identifiers": routing is on programId (base58); programLabel is the RPC guard string. Removed hardcoded labels

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 17 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Updated

File: openspec/changes/unify-instruction-parsing/proposal.md (hunk 20 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/proposal.md]: And yes, 15+ is fine for FSD. We could nest it inside something like decode-instruction/{system,token,token-2022,associated-token}. But I wouldn't mess with it right now, because I'm not completely sure where it should live (maybe even entities?)

File: app/features/decode-instruction-associated-token/lib/associated-token-parser.ts (hunk 21 above)
Comment (askov) [reply app/features/decode-instruction-associated-token/lib/associated-token-parser.ts:79]: Done

File: app/tx/__tests__/instruction-parser-contract.spec.ts (hunk 22 above)
Comment (askov) [reply app/tx/__tests__/instruction-parser-contract.spec.ts]: Added everything but MPL (can't -  no fromParsed, but it's still tested in metaplex-token-metadata-parser.spec.ts)

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 16 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:216]: Done

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 15 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:198]: Good call. added an explicit carve-out + scenario to the "Cards consume pre-parsed slice output" requirement

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 14 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Fixed

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 12 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Added a "known divergences", Weakened the requirement from "both satisfy the same superstruct"

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 11 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:143]: Looked into both suggestions and corrected the spec note:
- Readonly: not viable on data. KitInstruction.data is intentionally mutable Uint8Array, not ReadonlyUint8Array, because some kit parsers (lighthouse) require the mutable type
-  no-param-reassign: can be added as a guardrail and would catch ix.data = …, but not in-place element writes (ix.data[0] = …)/Object.assign,  which is the mutation that matters. It's not currently configured (this rule is good but won't help much)

File: openspec/changes/unify-instruction-parsing/proposal.md (hunk 10 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/proposal.md:36]: Added a plan to the proposal's "Follow-up areas" section.

File: openspec/changes/unify-instruction-parsing/proposal.md (hunk 9 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/proposal.md:34]: Added a plan to the proposal's "Follow-up areas" section.

File: openspec/changes/unify-instruction-parsing/proposal.md (hunk 8 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/proposal.md:27]: Changed to decode-*. But I'm still not sure what is the best place for these. maybe even entities. But feature layer seems to be a safe choice as a starter.

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 4 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:61]: Added it. dispatcher.canHandle(programId) is a cheap, parse-free predicate (pure registration lookup) so a caller can decide whether to route through a slice before attempting a decode

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 3 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md:57]: Yeah, the proposal mentioned this but the spec didn't, so I added it to the per-program parser requirement

File: openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md (hunk 1 above)
Comment (askov) [reply openspec/changes/unify-instruction-parsing/specs/instruction-parser/spec.md]: Clarified that

**Review by mikhd — APPROVED** [review]
🔥

File: app/components/inspector/InstructionsSection.tsx [hunk 23]
```diff
@@ -114,10 +110,16 @@ function InspectorInstructionCard({
     innerCards?: React.ReactNode[];
 }) {
     const { cluster, url } = useCluster();
+    const dispatcher = useInstructionParser();
 
     const programId = ix.programId;
     const programName = getProgramName(programId.toBase58(), cluster);
     const anchorProgram = useAnchorProgram(programId.toString(), url, cluster);
+    const parsedIx = React.useMemo(() => dispatcher.fromTransactionInstruction(ix), [dispatcher, ix]);
```
Comment (rogaldh) [inline app/components/inspector/InstructionsSection.tsx]: Could we import {useMemo} from React directly to match other files

**Review by rogaldh — APPROVED** [review]
LGTM

File: app/components/inspector/InstructionsSection.tsx (hunk 23 above)
Comment (askov) [reply app/components/inspector/InstructionsSection.tsx]: done
